python/ray/_private/log_monitor.py

```diff
@@ -503,6 +512,16 @@ def is_proc_alive(pid):
         args.logs_dir, gcs_pubsub.GcsPublisher(address=args.gcs_address), is_proc_alive
     )
 
+    # Set up handlers for graceful exit on SIGINT and SIGTERM
+    def graceful_exit(signum: int, _frame):
```

Comment:
Btw, just curious, but how does this work? Isn't the main thread already occupied by the loop? 

IIRC, python runs the interrupt on a main thread.

Reply:
Yeah, but apparently there is another abstraction between python and the underlying OS according to this: https://docs.python.org/3/library/signal.html#execution-of-python-signal-handlers

Not sure the exact code paths, but I believe it is OS -> some python code -> runs signal handler in main thread.